### ISSUE TITLE
Automatic pipeline for simple decoding of BIDS dataset

### ISSUE THREAD

**Issue by thomasbazeille** [nilearn/nilearn#2425]
As I worked on decoding several OpenfMRI datasets the past months, I wondered how much of the analysis can be made automatic to directly do inter or intrasubject decoding for BIDS datasets (or preprocessed with fmriprep) and I think many common cases are handlable.

I'm not sure whether this would really be useful for many users. I would just argue that when I realize how many datasets on OpenfMRI have no signal that I can decode with a principled pipeline, such an automatic tool could be a good incentive for people to try and decode their data as part of standard quality assesment / basic analysis.

The pipeline would includes :

0. Do a lot of checks (consistency of events/conditions across subjects/sessions, try to find files)

1. Fitting appropriate GLM for simple cases event-related or block-design (filtering and combining events based on user input, drawing design matrices with few default schemes, or using directly provided design matrices)

2. Storing outputs of the glm with appropriate label/run/subject information

3. Fit a decoder on appropriate data with appropriate CV scheme

4. Report the results (coef maps ? confusion matrix ?) 

I have most of the needed code base that could be quite factorized but API design would be a huge work and require a lot of feedback to be worth it. 

Would love to get your thoughts on this.

**Comment by bthirion** [comment]
>     1. Do a lot of checks (consistency of events/conditions across subjects/sessions, try to find files)

This is what you don't want to do. You should be able to rely on BIDS... otherwise, you have tons of boilerplate code.

**Comment by bthirion** [comment]
>     2\. Fitting appropriate GLM for simple cases event-related or block-design (filtering and combining events based on user input, drawing design matrices with few default schemes, or using directly provided design matrices)

I am worried by all the ORs that render the API super complex. We should target the most standard and straightforward use case, and extend it only if necessary.

**Comment by bthirion** [comment]
>     5\. Report the results (coef maps ? confusion matrix ?)

Yes all of that. basically, this is for free.

**Comment by bthirion** [comment]
Esssentially I like the idea. This is an excellent schowcase of what Nilearn should be able to do. Thx for the suggestion !

**Comment by GaelVaroquaux** [comment]
> You should be able to rely on BIDS...

+1 for code for BIDS, rather than for openeuro.

**Comment by GaelVaroquaux** [comment]
> +1 for code for BIDS, rather than for openeuro.

That was a stupid comment.

**Comment by jeromedockes** [comment]
> I have most of the needed code base that could be quite factorized but API design would be a huge work and require a lot of feedback to be worth it.

maybe put what you have in [`nilearn_sandbox`](https://github.com/nilearn/nilearn_sandbox) or some other public repo to help make the discussion concrete?

**Comment by thomasbazeille** [comment]
> maybe put what you have in nilearn_sandbox or some other public repo to help make the discussion concrete?
Ok I spent some time claryfing my ideas and code about this. Here is an example with I'd say 60-70% of the PR and some examples working very well on drago(but no tests) 
https://github.com/nilearn/nilearn_sandbox/tree/master/examples/bids_automatic_decoding

Please check it out and tell me what you think, I put below all opened questions that came during the factorization.

**Comment by thomasbazeille** [comment]
Main design problems :

1 Rely on BIDS modules ? 

Part 1 and 3 are really included in showcase code.
For discussions I included a bunch of datasets available on drago (if you don’t have access and want to add another one ; functions are pretty self-explanatory) but 
Part 1 is just mocked by : demo_datasets() and mocked_bids_fetcher()
because it entails a discussion which is a bit too advanced for now. (i.e. how do we handle BIDS datasets, add a dependency on pybids ? recode important steps ? )
Same for Part 3 you have to provide a save_location is automatically picked, I don’t know where would be a BIDS compliant place to store intermediate derivatives and if something in the BIDS suite could take care of that for us.

2 Factorization with FirstLevelModel estimation / Nilearn.stats in general

For now most of the important things for modelling happen in fit_design_matrix() 
Many function probably overlap with nilearn.stats functions, maybe FirstLevelModel heuristics. I didn’t have time to dig on how this could be more factorized. Maybe the paradigm design could be classes with a larger scope than decoding, i’m not sure. If you have any input or pointer, (@bthirion) happy to hear about it. 

Also I feel I already give much liberty to the end-user with a few parameters but maybe that would need some feedback

read_clean_paradigm(), loads events file, keep what is interesting for the user
def make_block_paradigm(), create a suitable paradigm structure to regress events block by block
def make_event_paradigm() create a suitable paradigm structure to regress events trial by trial
def _load_confounds(), load motion regressors from fmriprep confound.tsv file

3 About general design, and reusability for users

Not sure what the general design should be. Probably a class for pipeline, but not sure what it should inherit from. Also some smarter ways to have more factored code may be possible to find.
As this kind of pipeline is meant to be run only once per parameter set, it should be reusable with some kind of caching, or saving.
I guess a method to find the right files if they already exist or even the Decoder object with scores, and coefficients would also be needed. We don’t have often these problems in Nilearn for this it seems unavoidable.

4 About decoding
For Intra-subject I guess we should store one decoder per subject., what should we report then ?
The reporting part is untouched for now.
Should we have a transform method decoding any image ? Probably not suited, no transform() then ? 


Okay I think there is already way too much details for now so I’ll stop here. Just a few last important questions for later :

# Default heuristics should be carefuly crafted (masking, smoothing, filtering…)
# At which granularity to save derivatives ? for each trial, for each run ? for each sub ? (with according sub_labels.csv/ sub_runs.csv)
# Should we handle regression ? Probably not at first
# Should we give flexibility to the user on which column / combination of columns of the event_file to perform automatic fetching on ? At some point it’s almost easier for them to provide design matrices directly.
# Should I implement Inter and IntraSubject Pipeline together or separately ?
# Maybe include in Decoder an easy way to try Dummy decoder in the pipeline

**Comment by bthirion** [comment]
Some thoughts:
1. I'd rely on pybids. This could be an optional dependency if we want to use BIDS features
2. Sorry, I need to dig into that. Choice of additional parameters has to be motivated by actual cases.
3. I'd start with no specific class and rather see where it makes the code simpler. Don't see it as a silver bullet, but as a useful shortcut when the decoding problem is easily understood from the BIDS inputs (event files and list of classification targets among the trial_types)
4. To me such a model is not a transformer, but a classifier. But it has transformer(s) embedded.

**Comment by PeerHerholz** [comment]
@miykael and I worked on something related within [fmriflows](https://github.com/miykael/fmriflows/blob/master/notebooks/06_analysis_multivariate.ipynb), however using [pymvpa](http://www.pymvpa.org/index.html) as we wanted to include the `permutation based 2nd level analysis`. Happy to chat about it, the approach and experience.

**Comment by robbisg** [comment]
Hi everyone,

I've also worked on several ways to do semi-automatic pipelines for general decoding analyses (either across- and within-subject).
I wanted to share my design strategies, maybe it can be helpful, also to have some feedback!
My main goal was to run automatically different pipelines like in ```GridSearch```, without going into a serie of ```for``` loops, and try to explore different preprocessing pipeline (e.g. feature normalization in the whole dataset or at subject-level?), or different atlas
1. I use ```pybids``` in a ```DataLoader``` object to load data in ```pymvpa``` dataset format. I use ```pymvpa``` since I started with this package and I found its data structure very flexible to include different sets of labels (e.g. subject name for leave-one-subject-out analysis or different conditions) and different feature information (as masks). I load the whole dataset with all information I need and then I slice it for a particular analyses, this helps me to change easily the sklearn ```CrossValidation``` ```group``` attribute for example.
2. (but maybe also 3) I made a ```PreprocessingPipeline``` object that runs a set of ```Transformer``` objects, each of these has a ```save``` function to save intermediate results / derivatives. These ```Transformer``` are supposed to transform (of course) data to decoding functions. My idea was to include also ```GLM``` analysis to extract Beta images for example, but I have no time to do it now!
4. I used an ```Analysis``` object that specifies all the Decoding analyses (e.g. ```Searchlight```, ```RoiDecoding``` or ```TemporalDecoding```) is a kind of high-level ```nilearn``` wrapper in which I set a ```sklearn``` ```estimator``` ), a ```CrossValidation``` object, since intra-subject is a special case of Decoding in which you use a different attribute to split your dataset, and a ```fit``` function to run the analysis, and other attributes specific to the analysis (e.g. Searchlight radius). This function also has a ```save``` function that saves ```Searchlight``` maps, ```Decoding``` accuracies and other information etc, in a BIDS-like format.

Moreover I use an ```AnalysisPipeline``` object that runs ```DataLoader().fetch()```, ```PreprocessingPipeline().transform()```  data and ```Analysis().fit()```, using this object I can implement a way to run a vast set of analyses also using an ```AnalysisIterator``` and ```AnalysisConfiguration```.

I hope this helps!

If you'd like to see it, you can take a look at https://github.com/robbisg/pyitab .
Unfortunately it is badly documented, I'm sorry but I have no time to document it and to extensively test it. But if you want more infos, I will be very glad to talk about it.

**Comment by bthirion** [comment]
Thx for bringing all that forward. I think of these as long-term goals but we can start investigating what @robbisg and @PeerHerholz have done already. Ideally, we could chat about it.

**Comment by thomasbazeille** [comment]
Thanks a lot for the pointers @robbisg @PeerHerholz ! Would love to talk about it. @bthirion maybe we could take advantage of the sprint to have a quick chat ? 

Would a short 30mins session tomorrow around 9a EDT (3p CET)  be a good time ?

**Comment by robbisg** [comment]
If you want, I'd love to discuss about it, and 3pm (CET) LGTM.

**Comment by PeerHerholz** [comment]
I would be up for it, thx folks! “Where“ do we meet?

**Comment by GaelVaroquaux** [comment]
I'm not sure that I'll be able to meet. So one piece of advice: try to design individual components, such as functions to load a design matrix, rather  than a big integrated component. These will be easier to design in a future proof way, and will enable iterations 

⁣Sent from my phone. Please forgive typos and briefness.​

**Comment by bthirion** [comment]
In a Discord room, if that's OK for you ?

**Comment by thomasbazeille** [comment]
Everything sounds great to me ! 

So to recap : tomorrow Thursday 21th May at 9a EDT (3p CET) on the [sprint Discord](https://discord.gg/e4ZvxH7).

I'll add it on the calendar in case other people want to join.

**Comment by PeerHerholz** [comment]
Hi folks,

I have to do a few crowdcast tests for the scientific day of the dev days, hence I might be 10 minutes late. I'm truly sorry for any inconvenience this might cause.

**Comment by PeerHerholz** [comment]
Small update: @bthirion, @thomasbazeille and I will look into some examples from `fmriflows` and update asap.

**Comment by Remi-Gau** [comment]
In terms of choosing a dataset for this, some notes from an old core-dev meeting:

Add BIDS dataset with more recent fmriprep ?

- see https://github.com/nilearn/nilearn/pull/4198
- must have rest and task fMRI (ideally a task wiht event related design with several runs to decode on)
- Hao-Ting's downsampled datasets?
    - ds000228 (rest/movie watching, not uploaded yet)
    - [ds000017](https://zenodo.org/records/7901637#.ZFWJ15HMI4s) (task, on zenodo) 
- other possibility update examples we have to rename files / TSV where header where ncessary to fit our needs

**Comment by bthirion** [comment]
Also, having a smallish dataset is important for users. In the past, we downsampled data to 4mm to keep them small.
